### PR TITLE
add text recognition 0013 model

### PR DESCRIPTION
**Pull request by morkovka1337**
PR to support text-recognition-0013 model (demo + AC) which is aimed to replace existing text-recognition-0012 model trained in tensorflow.

### CODE REVIEW

**Comment by eaidova** [comment]
@morkovka1337 where is AC config file? I see only adapter code

File: tools/accuracy_checker/accuracy_checker/adapters/text_recognition.py [hunk 1]
```diff
@@ -40,7 +40,9 @@ def parameters(cls):
             ),
             'softmaxed_probabilities': BoolField(
                 optional=True, default=False, description="Indicator that model uses softmax for output layer "
-            )
+            ),
+            'logits_output': StringField(optional=True, description='Logits output layer name'),
+            'custom_label_map': StringField(optional=True, description='Label map')
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/adapters/text_recognition.py:45]: how it should be used?  is it path to file with label map or just string or list of tokens?
please note that as the rule, we use dictionary as label map

Comment (eaidova) [reply tools/accuracy_checker/accuracy_checker/adapters/text_recognition.py:45]: Please also provide new arguments in documentation (adapters/README.md)

Comment (morkovka1337) [reply tools/accuracy_checker/accuracy_checker/adapters/text_recognition.py:45]: The work is in progress now, this is a draft PR. It is planned to be a string, but I think, it wouldn't be difficult to pass a dict as label map.

File: demos/text_detection_demo/cpp/README.md [hunk 2]
```diff
@@ -6,6 +6,7 @@ The demo shows an example of using neural networks to detect and recognize print
 * `text-detection-0004`, which is a lightweight detection network for finding text.
 * `horizontal-text-detection-0001`, which is a detection network that works much faster than models above, but it is applicable to finding more or less horizontal text only.
 * `text-recognition-0012`, which is a recognition network for recognizing text.
+* `text-recognition-0013`, which is a recognition network for recognizing text.
```
Comment (eaidova) [inline demos/text_detection_demo/cpp/README.md:9]: is it added as replacement for 0012 or we will have 2 models?

File: models/intel/text-recognition-0013/description/text-recognition-0013.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+| ---------------------------------------------- | ------------------ |
+| Accuracy on the alphanumeric subset of ICDAR13 | 0.8739             |
+| Text location requirements                     | Tight aligned crop |
+| GFlops                                         | 1.488              |
+| MParams                                        | 5.567              |
+| Source framework                               | PyTorch            |
+
+## Inputs
+
+Shape: [1x1x32x120] - An input image in the format [BxCxHxW],
```
Comment (eaidova) [inline models/intel/text-recognition-0013/description/text-recognition-0013.md:24]:
```suggestion
Shape: `1, 1, 32, 120` - An input image in the format `B, C, H, W`,
```

File: models/intel/text-recognition-0013/description/text-recognition-0013.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+where:
+  - B - batch size
+  - C - number of channels
+  - H - image height
+  - W - image width
+
+Note that the source image should be tight aligned crop with detected text converted to grayscale.
+
+## Outputs
+The net outputs a blob with the shape [30, 1, 37] in the format [WxBxL],
```
Comment (eaidova) [inline models/intel/text-recognition-0013/description/text-recognition-0013.md:34]: as you have the issue with additional outputs, I think it will be useful if you add name for output in description (and for input too, for alignment)

Comment (eaidova) [inline models/intel/text-recognition-0013/description/text-recognition-0013.md:34]:
```suggestion
The net outputs a blob with the shape `30, 1, 37` in the format `W, B, L`,
```

**Comment by eaidova** [comment]
@morkovka1337, please do not forget to add model into index.md
```
text-recognition-0013 not in index.md file
```

File: demos/text_detection_demo/cpp/README.md (hunk 2 above)
Comment (morkovka1337) [reply demos/text_detection_demo/cpp/README.md:9]: For now, this would be an additiononal model.

File: models/intel/text-recognition-0013/description/text-recognition-0013.md (hunk 4 above)
Comment (morkovka1337) [reply models/intel/text-recognition-0013/description/text-recognition-0013.md]: Added info about inputs and outputs

File: demos/text_detection_demo/cpp/main.cpp [hunk 5]
```diff
@@ -276,7 +283,7 @@ int main(int argc, char *argv[]) {
 
             if (!FLAGS_no_show) {
                 cv::imshow("Press ESC or Q to exit", demo_image);
-                int key = cv::waitKey(1);
+                int key = cv::waitKey(0);
```
Comment (Wovchena) [inline demos/text_detection_demo/cpp/main.cpp:286]: Why zero now?

File: tools/accuracy_checker/accuracy_checker/adapters/text_recognition.py (hunk 1 above)
Comment (morkovka1337) [reply tools/accuracy_checker/accuracy_checker/adapters/text_recognition.py]: Type of label map changed to dict

**Review by eaidova — APPROVED** [review]

File: demos/text_detection_demo/cpp/main.cpp [hunk 6]
```diff
@@ -210,12 +213,16 @@ int main(int argc, char *argv[]) {
                 double conf = 1.0;
                 if (text_recognition.is_initialized()) {
                     auto blobs = text_recognition.Infer(cropped_text);
-                    auto output_shape = blobs.begin()->second->getTensorDesc().getDims();
+                    auto out_blob = blobs.begin()->second;
+                    if (FLAGS_tr_o_blb_nm != "")
+                        out_blob = blobs[FLAGS_tr_o_blb_nm];
+                    auto output_shape = out_blob->getTensorDesc().getDims();
+
                     if (output_shape[2] != kAlphabet.length()) {
```
Comment (Wovchena) [inline demos/text_detection_demo/cpp/main.cpp:221]: Please check `output_shape.size()` is at least 3. As I understand one of the outputs is of rank two and this results in reading garbage if wrong name is given.

File: models/intel/text-recognition-0013/description/text-recognition-0013.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+where:
+  - B - batch size
+  - C - number of channels
+  - H - image height
+  - W - image width
+
+Note that the source image should be tight aligned crop with detected text converted to grayscale.
+
+## Outputs
+The net outputs a blob `logits` with the shape `30, 1, 37` in the format `W, B, L`,
```
Comment (Wovchena) [inline models/intel/text-recognition-0013/description/text-recognition-0013.md:34]: Given demo has `-tr_o_blb_nm`, I think there should be multiple outputs listed.

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: demos/text_detection_demo/cpp/main.cpp (hunk 6 above)
Comment (morkovka1337) [reply demos/text_detection_demo/cpp/main.cpp]: Added check

File: demos/text_detection_demo/cpp/main.cpp [hunk 8]
```diff
@@ -210,12 +213,16 @@ int main(int argc, char *argv[]) {
                 double conf = 1.0;
                 if (text_recognition.is_initialized()) {
                     auto blobs = text_recognition.Infer(cropped_text);
-                    auto output_shape = blobs.begin()->second->getTensorDesc().getDims();
+                    auto out_blob = blobs.begin()->second;
+                    if (FLAGS_tr_o_blb_nm != "")
+                        out_blob = blobs[FLAGS_tr_o_blb_nm];
```
[the remark this answers is not in the document]
Comment (morkovka1337) [reply demos/text_detection_demo/cpp/main.cpp:222]: Added check that given output exist

File: models/intel/text-recognition-0013/description/text-recognition-0013.md (hunk 7 above)
Comment (morkovka1337) [reply models/intel/text-recognition-0013/description/text-recognition-0013.md:34]: All the other outputs are meaningless: they are created after model optimizer step and should not exist in the model. In other words, these are handing nodes.

Comment (Wovchena) [reply models/intel/text-recognition-0013/description/text-recognition-0013.md:34]: `input.yml` lists "targets".

File: demos/text_detection_demo/cpp/main.cpp [hunk 9]
```diff
@@ -210,12 +213,21 @@ int main(int argc, char *argv[]) {
...
                     auto blobs = text_recognition.Infer(cropped_text);
-                    auto output_shape = blobs.begin()->second->getTensorDesc().getDims();
-                    if (output_shape[2] != kAlphabet.length()) {
+                    auto out_blob = blobs.begin()->second;
+                    if (FLAGS_tr_o_blb_nm != ""){
+                        if (blobs.find(FLAGS_tr_o_blb_nm) == blobs.end()){
+                            std::string exception_msg = "The text recognition model does not have output " + std::string(FLAGS_tr_o_blb_nm);
+                            throw std::runtime_error(exception_msg);
+                        }
+                        out_blob = blobs[FLAGS_tr_o_blb_nm];
```
Comment (Wovchena) [inline demos/text_detection_demo/cpp/main.cpp:222]:
```suggestion
                    if (FLAGS_tr_o_blb_nm != "") {
                        const auto& iter = blobs.find(FLAGS_tr_o_blb_nm);
                        if (iter == blobs.end()) {
                            throw std::runtime_error("The text recognition model does not have output " + FLAGS_tr_o_blb_nm);
                        }
                        out_blob = iter->second;
```

File: demos/text_detection_demo/cpp/main.cpp [hunk 10]
```diff
@@ -210,12 +213,21 @@ int main(int argc, char *argv[]) {
...
+                    }
+                    auto output_shape = out_blob->getTensorDesc().getDims();
+
+                    if (output_shape.size() >=3 && output_shape[2] != kAlphabet.length()) {
```
Comment (Wovchena) [inline demos/text_detection_demo/cpp/main.cpp:226]:
```suggestion
                    if (output_shape.size() >= 3 && output_shape[2] != kAlphabet.length()) {
```

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: models/intel/text-recognition-0013/description/text-recognition-0013.md (hunk 7 above)
Comment (morkovka1337) [reply models/intel/text-recognition-0013/description/text-recognition-0013.md]: Oh, I've forgotten about this output since it is not used in accuracy checker nor in demo. Added info bout this output.

Comment (Wovchena) [reply models/intel/text-recognition-0013/description/text-recognition-0013.md]: Are you sure it has rank 3? For me IE prints 2.

**Review by Wovchena — APPROVED** [review]
